### PR TITLE
TypeScript migration.

### PR DESCRIPTION
**Pull request by mckoss**
@tomlarkworthy 

First pass of migration to TypeScript.  Tests are passing now.  I squashed the history for this branch.

Don't merge just yet - just wanted to get the diffs into PR.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,8 @@
 .DS_Store
+.idea
 bower_components/
 node_modules/
+lib/
+lib-test/
```
Comment (mckoss) [inline .gitignore]: No longer needed.

File: gulpfile.js [hunk 2]
```diff
@@ -28,21 +29,75 @@ var gutil = require('gulp-util');
...
+
+gulp.task('eslint', function() {
+  return gulp.src(JS_SOURCES)
     .pipe(eslint())
     .pipe(eslint.format())
     .pipe(eslint.failAfterError());
 });
 
-gulp.task('build', ['build-peg', 'browserify-bolt']);
+gulp.task('tslint', function() {
```
Comment (mckoss) [inline gulpfile.js]: Should have a 'lint' rule that does all.

File: gulpfile.js [hunk 3]
```diff
@@ -28,21 +29,75 @@ var gutil = require('gulp-util');
...
+  return gulp.src(TS_SOURCES)
+    .pipe(tslint())
+    .pipe(tslint.report('prose', {
+      emitError: false
+    }));
+});
+
+
+gulp.task('compile', ['build-peg'], function() {
```
Comment (mckoss) [inline gulpfile.js]: rename compile to ts-compile

File: src/.gitignore [hunk 4]
```diff
@@ -0,0 +1,8 @@
+# tools like IntilliJ put the .js in the same directory as .ts files, so ignore them
```
Comment (mckoss) [inline src/.gitignore]: Do we really need this?

File: src/ast.ts [hunk 5]
```diff
@@ -15,88 +15,51 @@
...
+export var gte = opGen('>=');
+export var ne = opGen('!=');
+export var and = opGen('&&');
+export var or = opGen('||');
+export var ternary = opGen('?:', 3);
+export var value = opGen('value', 1);
 
 var errors = {
-  typeMismatch: "Unexpected type: ",
+  typeMismatch: "Unexpected type: "
```
Comment (mckoss) [inline src/ast.ts]: revert

File: src/firebase-rest.ts [hunk 6]
```diff
@@ -60,10 +59,10 @@ util.methods(Client, {
     var options = {
       hostname: this.appName + '.' + FIREBASE_HOST,
       path: path + '.json',
-      method: opt.method,
+      method: opt.method
```
Comment (mckoss) [inline src/firebase-rest.ts]: revert

File: src/rules-generator.ts [hunk 7]
```diff
@@ -104,6 +96,8 @@ util.methods(Generator, {
     var name;
 
     for (name in paths) {
+      if (!paths.hasOwnProperty(name)) { continue; }
```
Comment (mckoss) [inline src/rules-generator.ts]: No inline blocks.

File: src/rules-generator.ts [hunk 8]
```diff
@@ -116,6 +110,8 @@ util.methods(Generator, {
     }
 
     for (var schemaName in this.symbols.schema) {
+      if (!this.symbols.schema.hasOwnProperty(schemaName)) { continue; }
```
Comment (mckoss) [inline src/rules-generator.ts]: Indent block

File: src/rules-generator.ts [hunk 9]
```diff
@@ -125,11 +121,13 @@ util.methods(Generator, {
     }
 
     for (var pathName in paths) {
+      if (!paths.hasOwnProperty(pathName)) { continue; }
```
Comment (mckoss) [inline src/rules-generator.ts]: Indent block

File: src/rules-generator.ts [hunk 10]
```diff
@@ -227,7 +225,7 @@ util.methods(Generator, {
   // Intermediate nodes can be "prop" or "$prop" values.
   createValidator: function(schemaName) {
     var schema = this.symbols.schema[schemaName];
-    var validator = {};
+    var validator: {$other?: any} = {}; // todo proper typing
```
Comment (mckoss) [inline src/rules-generator.ts]: I don't understand this type expression.

Comment (matjaz) [inline src/rules-generator.ts]: `validator` is an object with optional `$other` property of type `any`. this is called type literal in TS.

Comment (mckoss) [inline src/rules-generator.ts]: I don't think that's the correct type...

Comment (tomlarkworthy) [inline src/rules-generator.ts]: indeed, I somewhat myopically fixed a compile error, I am not sure what the type should really be, but its probably fundamental? You can type def a dictionary to give it a more meaningful name, without committing to a rigid structure (and then it makes refactoring over the validator type easy)
e.g. type Validator = {[key:name]:string}

**Comment by mckoss** [comment]
Merged via b70f3d3645e2d446b9c6b350c9aef22f225558f0

**Pull request closed without merging**